File: examples/javascript/smart-account.js
/**
 * NOVIS Smart Account Example
 * 
 * Create and manage AI agent smart accounts with spending limits.
 * 
 * Usage:
 *   PRIVATE_KEY=0x... node smart-account.js create <daily_limit>
 *   PRIVATE_KEY=0x... node smart-account.js fund <account_address> <amount>
 *   PRIVATE_KEY=0x... node smart-account.js info <account_address>
 */

import { ethers } from 'ethers';

// Config
const NOVIS_TOKEN = '0x1fb5e1C0c3DEc8da595E531b31C7B30c540E6B85';
const SMART_ACCOUNT_FACTORY = '0x4b84E3a0D640c9139426f55204Fb34dB9B1123EA';
const RPC_URL = 'https://mainnet.base.org';

// ABIs
const TOKEN_ABI = [
  'function transfer(address to, uint256 amount) returns (bool)',
  'function balanceOf(address account) view returns (uint256)'
];

const FACTORY_ABI = [
  'function createAccount(address owner, uint256 dailyLimit) returns (address)',
  'function getAccount(address owner) view returns (address)',
  'event AccountCreated(address indexed owner, address indexed account, uint256 dailyLimit)'
];

const SMART_ACCOUNT_ABI = [
  'function owner() view returns (address)',
  'function dailyLimit() view returns (uint256)',
  'function spentToday() view returns (uint256)',
  'function lastSpendDate() view returns (uint256)',
  'function execute(address to, uint256 value, bytes data) returns (bytes)',
  'function setDailyLimit(uint256 newLimit)'
];

async function createSmartAccount(wallet, dailyLimit) {
  const factory = new ethers.Contract(SMART_ACCOUNT_FACTORY, FACTORY_ABI, wallet);

  console.log('Creating smart account...');
  console.log(`  Owner: ${wallet.address}`);
  console.log(`  Daily Limit: ${dailyLimit} NOVIS`);

  const limitWei = ethers.parseEther(dailyLimit);
  const tx = await factory.createAccount(wallet.address, limitWei);
  console.log(`TX Hash: ${tx.hash}`);

  const receipt = await tx.wait();

  // Parse account address from event
  let accountAddress = null;
  for (const log of receipt.logs) {
    try {
      const parsed = factory.interface.parseLog(log);
      if (parsed?.name === 'AccountCreated') {
        accountAddress = parsed.args.account;
        break;
      }
    } catch {}
  }

  if (accountAddress) {
    console.log(`\n✅ Smart Account created!`);
    console.log(`   Address: ${accountAddress}`);
    console.log(`\nNext steps:`);
    console.log(`  1. Fund it: node smart-account.js fund ${accountAddress} 100`);
    console.log(`  2. Use it for AI agent transactions`);
  } else {
    // Try to get existing account
    accountAddress = await factory.getAccount(wallet.address);
    console.log(`\n⚠️ Account may already exist: ${accountAddress}`);
  }
}

async function fundSmartAccount(wallet, accountAddress, amount) {
  const token = new ethers.Contract(NOVIS_TOKEN, TOKEN_ABI, wallet);

  console.log('Funding smart account...');
  console.log(`  Account: ${accountAddress}`);
  console.log(`  Amount: ${amount} NOVIS`);

  const amountWei = ethers.parseEther(amount);
  const tx = await token.transfer(accountAddress, amountWei);
  console.log(`TX Hash: ${tx.hash}`);

  await tx.wait();
  console.log(`\n✅ Funded successfully!`);

  // Check new balance
  const balance = await token.balanceOf(accountAddress);
  console.log(`Account balance: ${ethers.formatEther(balance)} NOVIS`);
}

async function getAccountInfo(provider, accountAddress) {
  const token = new ethers.Contract(NOVIS_TOKEN, TOKEN_ABI, provider);
  const account = new ethers.Contract(accountAddress, SMART_ACCOUNT_ABI, provider);

  console.log('Smart Account Info');
  console.log('='.repeat(50));
  console.log(`Address: ${accountAddress}`);

  try {
    const owner = await account.owner();
    const dailyLimit = await account.dailyLimit();
    const spentToday = await account.spentToday();
    const balance = await token.balanceOf(accountAddress);

    console.log(`Owner: ${owner}`);
    console.log(`Balance: ${ethers.formatEther(balance)} NOVIS`);
    console.log(`Daily Limit: ${ethers.formatEther(dailyLimit)} NOVIS`);
    console.log(`Spent Today: ${ethers.formatEther(spentToday)} NOVIS`);
    console.log(`Remaining: ${ethers.formatEther(dailyLimit - spentToday)} NOVIS`);
  } catch (e) {
    console.log(`\n⚠️ Could not read account info. Is this a valid smart account?`);
    console.log(`   Error: ${e.message}`);
  }
}

async function main() {
  const command = process.argv[2];

  if (!command || !['create', 'fund', 'info'].includes(command)) {
    console.log('NOVIS Smart Account Manager');
    console.log('='.repeat(50));
    console.log('Usage:');
    console.log('  node smart-account.js create <daily_limit>');
    console.log('  node smart-account.js fund <account_address> <amount>');
    console.log('  node smart-account.js info <account_address>');
    console.log('');
    console.log('Examples:');
    console.log('  node smart-account.js create 100        # Create with 100 NOVIS/day limit');
    console.log('  node smart-account.js fund 0x... 50     # Fund with 50 NOVIS');
    console.log('  node smart-account.js info 0x...        # Get account info');
    process.exit(1);
  }

  const provider = new ethers.JsonRpcProvider(RPC_URL);

  if (command === 'info') {
    const accountAddress = process.argv[3];
    if (!accountAddress) {
      console.error('Error: account_address required');
      process.exit(1);
    }
    await getAccountInfo(provider, accountAddress);
    return;
  }

  // Other commands need private key
  if (!process.env.PRIVATE_KEY) {
    console.error('Error: PRIVATE_KEY environment variable required');
    process.exit(1);
  }

  const wallet = new ethers.Wallet(process.env.PRIVATE_KEY, provider);
  console.log(`Wallet: ${wallet.address}\n`);

  switch (command) {
    case 'create':
      const dailyLimit = process.argv[3] || '100';
      await createSmartAccount(wallet, dailyLimit);
      break;

    case 'fund':
      const accountAddress = process.argv[3];
      const amount = process.argv[4];
      if (!accountAddress || !amount) {
        console.error('Error: account_address and amount required');
        process.exit(1);
      }
      await fundSmartAccount(wallet, accountAddress, amount);
      break;
  }
}

main().catch(console.error);

File: examples/python/smart_account.py
"""
NOVIS Smart Account Example (Python)

Create and manage AI agent smart accounts with spending limits.

Usage:
    PRIVATE_KEY=0x... python smart_account.py create <daily_limit>
    PRIVATE_KEY=0x... python smart_account.py fund <account_address> <amount>
    python smart_account.py info <account_address>
"""

import os
import sys
from web3 import Web3
from eth_account import Account

# Config
NOVIS_TOKEN = '0x1fb5e1C0c3DEc8da595E531b31C7B30c540E6B85'
SMART_ACCOUNT_FACTORY = '0x4b84E3a0D640c9139426f55204Fb34dB9B1123EA'
RPC_URL = 'https://mainnet.base.org'
CHAIN_ID = 8453

# ABIs
TOKEN_ABI = [
    {"name": "transfer", "type": "function",
     "inputs": [{"name": "to", "type": "address"}, {"name": "amount", "type": "uint256"}],
     "outputs": [{"type": "bool"}]},
    {"name": "balanceOf", "type": "function", "stateMutability": "view",
     "inputs": [{"name": "account", "type": "address"}],
     "outputs": [{"type": "uint256"}]}
]

FACTORY_ABI = [
    {"name": "createAccount", "type": "function",
     "inputs": [
         {"name": "owner", "type": "address"},
         {"name": "dailyLimit", "type": "uint256"}
     ],
     "outputs": [{"name": "", "type": "address"}]},
    {"name": "getAccount", "type": "function", "stateMutability": "view",
     "inputs": [{"name": "owner", "type": "address"}],
     "outputs": [{"name": "", "type": "address"}]}
]

SMART_ACCOUNT_ABI = [
    {"name": "owner", "type": "function", "stateMutability": "view",
     "inputs": [],
     "outputs": [{"type": "address"}]},
    {"name": "dailyLimit", "type": "function", "stateMutability": "view",
     "inputs": [],
     "outputs": [{"type": "uint256"}]},
    {"name": "spentToday", "type": "function", "stateMutability": "view",
     "inputs": [],
     "outputs": [{"type": "uint256"}]},
    {"name": "lastSpendDate", "type": "function", "stateMutability": "view",
     "inputs": [],
     "outputs": [{"type": "uint256"}]}
]


def get_web3():
    return Web3(Web3.HTTPProvider(RPC_URL))


def get_account():
    private_key = os.environ.get('PRIVATE_KEY')
    if not private_key:
        print('Error: PRIVATE_KEY environment variable required')
        sys.exit(1)
    return Account.from_key(private_key)


def create_smart_account(daily_limit: str):
    w3 = get_web3()
    account = get_account()
    factory = w3.eth.contract(address=SMART_ACCOUNT_FACTORY, abi=FACTORY_ABI)

    print('Creating smart account...')
    print(f'  Owner: {account.address}')
    print(f'  Daily Limit: {daily_limit} NOVIS')

    limit_wei = w3.to_wei(float(daily_limit), 'ether')

    nonce = w3.eth.get_transaction_count(account.address)
    tx = factory.functions.createAccount(
        account.address,
        limit_wei
    ).build_transaction({
        'from': account.address,
        'nonce': nonce,
        'gas': 500000,
        'gasPrice': w3.eth.gas_price,
        'chainId': CHAIN_ID
    })

    signed = account.sign_transaction(tx)
    tx_hash = w3.eth.send_raw_transaction(signed.raw_transaction)
    print(f'TX Hash: {tx_hash.hex()}')

    receipt = w3.eth.wait_for_transaction_receipt(tx_hash)

    # Get account address
    smart_account_address = factory.functions.getAccount(account.address).call()

    print(f'\n✅ Smart Account created!')
    print(f'   Address: {smart_account_address}')
    print(f'\nNext steps:')
    print(f'  1. Fund it: python smart_account.py fund {smart_account_address} 100')
    print(f'  2. Use it for AI agent transactions')


def fund_smart_account(account_address: str, amount: str):
    w3 = get_web3()
    account = get_account()
    token = w3.eth.contract(address=NOVIS_TOKEN, abi=TOKEN_ABI)

    print('Funding smart account...')
    print(f'  Account: {account_address}')
    print(f'  Amount: {amount} NOVIS')

    amount_wei = w3.to_wei(float(amount), 'ether')

    nonce = w3.eth.get_transaction_count(account.address)
    tx = token.functions.transfer(
        account_address,
        amount_wei
    ).build_transaction({
        'from': account.address,
        'nonce': nonce,
        'gas': 100000,
        'gasPrice': w3.eth.gas_price,
        'chainId': CHAIN_ID
    })

    signed = account.sign_transaction(tx)
    tx_hash = w3.eth.send_raw_transaction(signed.raw_transaction)
    print(f'TX Hash: {tx_hash.hex()}')

    w3.eth.wait_for_transaction_receipt(tx_hash)
    print(f'\n✅ Funded successfully!')

    # Check new balance
    balance = token.functions.balanceOf(account_address).call()
    print(f'Account balance: {w3.from_wei(balance, "ether")} NOVIS')


def get_account_info(account_address: str):
    w3 = get_web3()
    token = w3.eth.contract(address=NOVIS_TOKEN, abi=TOKEN_ABI)
    smart_account = w3.eth.contract(address=account_address, abi=SMART_ACCOUNT_ABI)

    print('Smart Account Info')
    print('=' * 50)
    print(f'Address: {account_address}')

    try:
        owner = smart_account.functions.owner().call()
        daily_limit = smart_account.functions.dailyLimit().call()
        spent_today = smart_account.functions.spentToday().call()
        balance = token.functions.balanceOf(account_address).call()

        print(f'Owner: {owner}')
        print(f'Balance: {w3.from_wei(balance, "ether")} NOVIS')
        print(f'Daily Limit: {w3.from_wei(daily_limit, "ether")} NOVIS')
        print(f'Spent Today: {w3.from_wei(spent_today, "ether")} NOVIS')
        print(f'Remaining: {w3.from_wei(daily_limit - spent_today, "ether")} NOVIS')
    except Exception as e:
        print(f'\n⚠️ Could not read account info. Is this a valid smart account?')
        print(f'   Error: {e}')


def main():
    if len(sys.argv) < 2:
        print('NOVIS Smart Account Manager')
        print('=' * 50)
        print('Usage:')
        print('  python smart_account.py create <daily_limit>')
        print('  python smart_account.py fund <account_address> <amount>')
        print('  python smart_account.py info <account_address>')
        print('')
        print('Examples:')
        print('  python smart_account.py create 100        # Create with 100 NOVIS/day limit')
        print('  python smart_account.py fund 0x... 50     # Fund with 50 NOVIS')
        print('  python smart_account.py info 0x...        # Get account info')
        sys.exit(1)

    command = sys.argv[1]

    if command == 'create':
        daily_limit = sys.argv[2] if len(sys.argv) > 2 else '100'
        create_smart_account(daily_limit)

    elif command == 'fund':
        if len(sys.argv) < 4:
            print('Error: account_address and amount required')
            sys.exit(1)
        fund_smart_account(sys.argv[2], sys.argv[3])

    elif command == 'info':
        if len(sys.argv) < 3:
            print('Error: account_address required')
            sys.exit(1)
        get_account_info(sys.argv[2])

    else:
        print(f'Unknown command: {command}')
        sys.exit(1)


if __name__ == '__main__':
    main()
